### ISSUE TITLE
Godot and the risks of nontrivial relocatability

### ISSUE THREAD

**Issue by Ivorforce** [godotengine/godot#100509]
### Tested versions

All versions are affected.

### System information

Unimportant.

### Issue description

Godot currently has many hidden assumptions about trivial relocatability. Let's gather knowledge about relocations in this issue, and see how we can best handle them

## About Relocations 

_Relocation_ of an object means moving it to a different position in memory.

Relocation may commonly be implemented by `memcpy` / `memmove` with subsequent non-destructing deallocation of the old memory. Calls to `realloc` _may_ trigger the same operation, if a new array is allocated.

This brings us to _trivial relocation_:
- An object is _trivially relocatable_ if the above operation is _correct_.
- An object is _not trivially relocatable_ if the above operation is incorrect, such as if:
    - The object has a pointer to itself.
    - The object uses relative pointers.
    - Other example can be found [here](https://open-std.org/JTC1/SC22/WG21/docs/papers/2020/p1144r5.html#non-trivial-samples).

A violation of nontrivial relocatability is undefined behavior, and will likely lead to a crash.

The verb 'relocate' was introduced rather recently, and didn't make it even into C++20: https://open-std.org/JTC1/SC22/WG21/docs/papers/2020/p1144r5.html

To have safe relocation dynamics, we need to either avoid nontrivial relocation entirely (inefficient) or introduce relocation structures into the Godot codebase.

## How to handle non-trivial relocatability

If an object is not trivially relocatable, it needs to be moved using _relocation semantics_ instead. For example (pseudocode):

```c++
if constexpr (std::is_trivially_relocatable<T>) {
    realloc(array, new_size);
}
else {
    // Not trivially relocatable
    T *new_array = alloc(new_size);
    for (int i = 0; i < size; i++) {
        new_array[i] = relocate(array[i]);
    }
}
```

In the common case, `relocate` may be implemented as `std::move`.

## Examples of code violating nontrivial relocatability

- `CowData`, through use of `realloc`, affecting `Vector`.
- `LocalVector`, through use of `realloc`.

I'm sure there are more examples; please comment any that you know of so we can gather them here :)

## Examples of nontrivially relocatable types

I don't know any, please comment if you can think of any.
GDExtension users may have their own nontrivially relocatable types.

## Practical Proposal to move forwards

I don't think Godot has many nontrivially relocatable types. It is questionable whether we want to sacrifice time to implement relocation-safe versions of code, especially since the trivially relocatable path should not be changed, since it is likely to be faster and should be accelerated.

Instead, I propose we document such limitations, and statically safeguard incompatible types such as `CowData` against nontrivially relocatable element types. We can remove this limitation if needed later.

### Steps to reproduce

N/A

### Minimal reproduction project (MRP)

N/A

**Comment by hpvb** [comment]
Yeah, I think the first order of business is to ensure we do static assertions to reduce footgun potential! That would be "free" and would at least help us figure out if this is something worth caring about.

**Comment by Ivorforce** [comment]
We discussed this problem in a core meeting.

The verdict was the following:

- Just assume **everything** is trivially relocatable.
- If we find something that isn't trivially relocatable, then we should make it trivially relocatable.
    - If we find something that _needs to be_ non trivially relocatable, then reconsider.
- By assuming trivial relocatability for everything, we unlock more optimizations (such as uses of `memmove` and `realloc`).

Not sure whether to leave the Issue open, or just close it (and reopen if it becomes relevant).
